flit/validate.py

```diff
@@ -149,37 +153,88 @@ def _valid_version_specifier(s):
-    problems = []
-    for c in clauses:
-        # TODO: validate parentheses properly. They're allowed by PEP 508.
-        parts = MARKER_OP.split(c.strip('()'))
-        if len(parts) != 3:
-            problems.append("Invalid expression in environment marker: {!r}".format(c))
-            continue
-        l, op, r = parts
+
+    def reduce_expression():
+        # EXP OPS EXP -> EXP
+        stk.pop()
+        stk.pop()
+        stk.pop()
+        stk.append("EXP")
+
+    def verify_statement(l, op, r):
         for var in (l.strip(), r.strip()):
             if var[:1] in {'"', "'"}:
                 if len(var) < 2 or var[-1:] != var[:1]:
                     problems.append("Invalid string in environment marker: {}".format(var))
             elif var not in MARKER_VARS:
                 problems.append("Invalid variable name in environment marker: {!r}".format(var))
+
+    em = em.replace("(", " ( ").replace(")", " ) ").split(" ")
+    token = list('(') + list(filter(lambda k: k != "", em)) + list(')')
+    problems = []
+    idx = 0
+    stk = []
+    try:
+        while idx < len(token):
+            if token[idx] == '(':
+                stk.append('(')
+            elif token[idx] == ')':
+                if '(' not in stk:
+                    raise Exception("Invalid expression. incorrect parentheses")
+                while stk[-1] != '(' and len(stk) > 0:
+                    if stk[-1] == "EXP":
+                        if stk[-2] in {"and", "or"} and stk[-3] in {"and", "or"}:
+                            raise Exception("Invalid expression \"{} {}\"".format(stk[-3], stk[-2]))
+                        if len(stk) > 3 and stk[-2] in {"and", "or"}:
+                            reduce_expression()
+                        elif stk[-2] == '(':
+                            stk.pop()
+                        else:
+                            raise Exception("Invalid expression \"{}\"".format(stk[-2]))
+                    else:
+                        raise Exception("Invalid expression \"{}\"".format(stk[-1]))
+                stk.pop()
+                stk.append("EXP")
+            else:
+                if idx > 1:
+                    if MARKER_OP.match(stk[-1]):
+                        if stk[-2] == "EXP":
+                            raise Exception("Invalid expression")
+                        l, op, r = token[idx - 2:idx + 1]
+                        verify_statement(l, op, r)
+                        stk.append(token[idx])
+                        reduce_expression()
+                        if len(stk) > 1 and stk[-1] in {"and", "or"}:
+                            reduce_expression()
+                    else:
+                        stk.append(token[idx])
+                else:
+                    stk.append(token[idx])
+            idx += 1
+        if len(stk) != 1 or stk[-1] != "EXP":
+            problems.append("Invalid environment markers syntax")
+    except Exception as e:
```

Comment:
I don't like this structure - it could easily obscure a bug in the code by catching an overly general exception, and it means all the interesting code is indented a level more than it needs to be.

If you want to jump out of parsing as soon as a problem is detected, I think it's OK to use an early return.